gather_town_pipeline.py

```diff
@@ -0,0 +1,143 @@
+          .sort_values(['name', 'timestamp'])
+          .reset_index(drop=True)
+          )
+
+    checks['timestamp'] = checks['timestamp'] - pd.Timedelta('7 hours')
+    checks['elapsed_min'] = checks.groupby('name')['timestamp'].diff(1).dt.total_seconds().div(60).round(2)
+    checks['first_visit'] = False
+    checks.loc[checks['elapsed_min'].isnull(), 'first_visit'] = True
+
+    checks['new_session'] = False
+    checks.loc[checks['elapsed_min'] > 30, 'new_session'] = True
+    checks['session_no'] = checks.groupby('name')['new_session'].cumsum().fillna(0).astype(int)
+
+    checks['busy'] = checks['busy'].fillna(0).astype(int)
+
+    checks['x_coord'] = checks['location'].str['x']
+    checks['y_coord'] = checks['location'].str['y']
+    checks['map'] = checks['location'].str['map']
+
+    checks['lw_team'] = checks['name'].isin(['Raemon', 'Habryka', 'Ruby', 'Ben Pace', 'jimrandomh'])
+
+    for col in ['audio', 'video', 'blocked']:
+        checks[col] = checks[col].fillna(False)
+
+    checks_output_cols = ['timestamp', 'name', 'busy', 'audio', 'video', 'blocked', 'elapsed_min', 'first_visit',
+                          'new_session', 'session_no', 'lw_team']
+
+    return checks[checks_output_cols]
+
+
+def generate_presence_table(checks):
+    presence = (checks
+                .groupby(['timestamp'])['name']
+                .agg(['nunique', lambda x: x.tolist()])
+                .fillna(0)
+                .sort_index()
+                )
+
+    #     presence['total_present'] = presence[[F.sum(axis=1)
+    presence.columns = ['num_present', 'names']
+
+    return presence
+
+
+def generate_sessions_table(checks):
+
+    presence = generate_presence_table(checks)
+
+    sessions = (checks
+        .merge(presence, on='timestamp')
+        .sort_values('timestamp')
+        .groupby(['name', 'session_no'])
+        .agg({
+        'audio': 'size',
+        'elapsed_min': lambda x: x.iloc[1:].max(),
+        'timestamp': ['min', 'max'],
+        'first_visit': np.any,
+        'lw_team': np.any,
+        'num_present': [lambda x: x.iloc[0]<2, lambda x: x.iloc[-1]<2, lambda x: (x>1).mean().round(2)],
+        'names': lambda x: list(np.unique(x.sum()))
+    })
+    )
+
+    sessions.columns = ['num_checks', 'max_gap', 'start_time', 'end_time', 'first_visit', 'lw_team', 'alone_at_start',
+                        'alone_at_end', 'percent_accompanied', 'concurrent_visitors']
+    sessions['approx_duration'] = (sessions['end_time'] - sessions['start_time']).dt.total_seconds().div(60).round(2)
+    sessions = sessions.reset_index()
+    sessions.apply(lambda x: x['concurrent_visitors'].remove(x['name']), axis=1) #.sample(10)
+    sessions['concurrent_visitors'] = sessions['concurrent_visitors'].astype(str)
+
+    return sessions
+
+
+def generate_users_table(sessions):
+
+    user_stats = (sessions
+                  .groupby('name')
+                  .agg({
+        'session_no': 'size',
```

Comment:
I am not sure whether I'm supposed to do something differently...

Reply:
I think python is idiosyncratic in that it's programmers prefer to do this
```
foobar(baz, quux,
       zip, fee)
```
instead of
```
foobar(
    baz,
    quux,
    zip,
    fee
)
```
which I think works well for a lot of cases, that makes the language feel very compact and readable, but only in the cases where it works. And when we when it doesn't work, as in here, it's kind of awful.

Now if you told me you preferred the above style, I wouldn't object at all, it absolutely could be the right choice in the end. But of now as a time were we decide on a style, I'd say that this is a good case for the second.